Jenkinsfile

```diff
@@ -0,0 +1,307 @@
+// if ( env.ghprbGhRepository != null ) {
+//   // but, if this PR is from the tests repo, use the Jenkins checked out
+//   // source/branch of that instead, in case the PR is updating the test code.
+//   if ( env.ghprbGhRepository == env.ci_repo_name ) {
+//     testcode_rootdir = ""
+//   }
+// }
+        //dir("${testcode_rootdir}") {
+          script {
+            sh "pwd"
+            sh "ls"
+            testmatrix = readYaml file: "${testmatrix_file}"
+            echo "testmatrix looks like ${testmatrix}"
+         // }
+        }
+      }
+    }
+
+    // Run the CI on the primary distro (note, that does not mean this is our
+    // favoured distro, it just means we needed to pick one as the primary
+    // smoke test). If the primary distro does not pass, the rest of the distro
+    // tests will be skipped by Jenkins.
+    stage('Ubuntu18.04') {
+      when {
+        expression { SKIPTESTS != 'true' }
+      }
+      agent { label "master" }
+      steps {
+        echo "Am running ${STAGE_NAME}"
+        testrunner ( STAGE_NAME )
+      }
+    }
+
+    // If we have passed all previous steps, finally we can parallel run the
+    // rest of the distro checks.
+    stage('Run main jobs') {
+      when {
+        expression { SKIPTESTS != 'true' }
+      }
+      parallel {
+        stage('Fedora') {
+          agent { label "master" }
+          steps {
+            echo "Am running ${STAGE_NAME}"
+            testrunner ( STAGE_NAME )
+          }
+        }
+        stage('Centos') {
+          agent { label "master" }
+          steps {
+            echo "Am running ${STAGE_NAME}"
+            testrunner ( STAGE_NAME )
+          }
+        }
+      }
+    }
+  }
+}
+
+
+// Run our tests on the 'distroName' distro provided.
+// Use that name to look in the testmatrix to work out which tests
+// are to be run on this repo/distro combination.
+
+def testrunner(distroName) {
+  echo "Running tests for distro (${distroName})"
+
+  // Check here so we can mark a 'stage' as skipped
+  if ( env.ghprbGhRepository == null ) {
+    stage('skipping') {
+      echo "Skipping tests, ghprbGhRepository variable not set"
+      // This only returns out of this stage...
+      return
+    }
+  }
+
+  // And now really return from this func
+  if ( env.ghprbGhRepository == null ) {
+    return
+  }
+
+  echo "testmatrix looks like ${testmatrix}"
+  testmap = testmatrix[env.ghprbGhRepository]
+  echo "testmap[repo] looks like ${testmap}"
+  testmap = testmatrix[env.ghprbGhRepository][distroName]
+  echo "testmap[repo][distro] looks like ${testmap}"
+
+  echo "testmap['docker'] looks like ${testmap['docker']}"
+
+  if ( testmap['docker'] == true ) {
+    stage('docker') {
+      dir("${GOPATH}/src/${env.ghprbGhRepository}") {
+        echo "Running docker tests"
+        sh "pwd"
+        sh "echo make docker"
+      }
+    }
+  } else {
+    echo "Docker tests disabled for ${distroName}/${env.ghprbGhRepository}"
+  }
+}
+
+@NonCPS
```

Comment:
What's this bit of magic do?

Reply:
It's a security scoping thing - 'standard' Jenkins groovy pipeline stuff, but yes, it will gain a comment.